Relative source syntax sugar

jkoritzinsky opened the pull request:
Syntax sugar for RelativeSource bindings. Fixes #590.

You can now use the following shorthand for RelativeSource bindings:


Shorthand | Implementation
--------|---------
`$self` | `Mode = Self`
`$parent` | `Mode = FindAncestor; AncestorLevel = 1`
`$parent[Level]` | `Mode = FindAncestor; AncestorLevel = Level +1`
`$parent[ns:Type]` | `Mode = FindAncestor; AncestorType = ns:Type`
`$parent[ns:Type; Level]` | `Mode = FindAncestor; AncestorType = ns:Type; AncestorLevel = Level + 1`

jkoritzinsky committed:
Implement RelativeSource binding syntax sugar except for parsing AncestorType.

jkoritzinsky committed:
Update XUnit to a released version.

jkoritzinsky committed:
Implement syntax sugar for RelativeSource (not including setting a specific AncestorType)

grokys commented on src/Markup/Avalonia.Markup.Xaml/Data/Binding.cs:
```diff
@@ -187,6 +191,69 @@ private static PathInfo ParsePath(string path)
+                }
+                else if(relativeSourceMode == "parent")
+                {
+                    relativeSource.Mode = RelativeSourceMode.FindAncestor;
```
This probably isn't what we want here, `RelativeSource` works on the visual tree, so:

```
<ContentControl>
  <TextBlock Text="$parent.Foo"/>
<ContentControl>
```

Will bind to the `ContentPresenter` in the `ContentControl` rather than the `ContentControl`. This is because we follow WPF's semantics for `RelativeSource` and that works on the visual rather than logical tree. This is OK in WPF because you have to specify the type of the ancestor, but here one would intuitively expect it to work on the logical tree.

grokys commented on src/Markup/Avalonia.Markup.Xaml/Data/Binding.cs:
```diff
@@ -195,6 +262,12 @@ private static PathInfo ParsePath(string path)
+        private static Type LookupAncestorType(string ancestorTypeName)
+        {
+            //TODO: What is our syntax for type lookup here?
+            throw new NotImplementedException();
```
Would simply `$parent[ns:ControlType]` work here? To do this I think we'd need to make `Binding` a `MarkupExtension` so we have access to the XAML context. This is how it is in WPF but we couldn't do it previously because of how OmniXAML worked(https://github.com/OmniGUI/OmniXAML/issues/74) but we should be able to do it now we're using Portable.Xaml.

grokys reviewed (commented):
Left a few comments. If you agree, I wonder if maybe we should do this in a couple of separate PRs:

- Make `Binding` a markup extension
- Add a `Tree=Logical` parameter to `RelativeSource`

Aside: We can probably simplify things a bit in our code here because any class can be used as a markup extension in XAML (https://github.com/cwensley/Portable.Xaml/issues/65) so we can probably merge a few classes...

jkoritzinsky commented:
I'm ok with making Binding a markup extension,. My main curiosity right now with that is the separation between `Binding` and `XamlBinding`. would we combine those two classes as well?

grokys commented:
Yeah I think those 2 classes could be merged. I did something similar when I went from `StyleResource` to `DynamicResource`/`StaticResource`.

jkoritzinsky committed:
Move path parsing into the binding extension. Our path shorthands will only work from XAML now. Now we can look up types.

jkoritzinsky committed:
If AncestorType == null, track ancestor on logical tree instead of visual tree.

jkoritzinsky commented:
@grokys I've moved the path shorthand parsing to the `BindingExtension` class.

I also changed the `FindAncestor` lookup when `AncestorType == null` to lookup on the logical tree instead, as well as fix some bugs in `ControlLocator`.

jkoritzinsky committed:
Fixed bugs.

jkoritzinsky committed:
Remove invalid param doc tag

grokys commented on src/Markup/Avalonia.Markup/ControlLocator.cs:
```diff
@@ -100,5 +100,42 @@ public static IObservable<IControl> Track(IControl relativeTo, Type ancestorType
+
+            var detached = Observable.FromEventPattern<LogicalTreeAttachmentEventArgs>(
+                x => relativeTo.DetachedFromLogicalTree += x,
+                x => relativeTo.DetachedFromLogicalTree += x)
```
I think this should be `-=`.

jkoritzinsky replied:
Good catch! I'll fix that.

grokys reviewed (commented):
This is great @jkoritzinsky. Not had chance for a full review yet, but found something that doesn't look right.

jkoritzinsky committed:
Fix unsubscription.

grokys committed:
Added more tests.

One failing.

grokys commented:
I'm a little unsure about switching between logical and visual trees depending on whether a type is specified. For example, I'd expect these two snippets to do the same:

```
<ContentControl>
  <TextBlock Tag="{Binding $parent[0]}"/>
</ContentControl>
```

```
<ContentControl>
  <TextBlock Tag="{Binding $parent[Control;0]}"/>
</ContentControl>
```

But currently they won't: the first will bind to `ContentControl` and the second to `ContentPresenter`. This strikes me as a potential source of confusion...

jkoritzinsky commented:
So do we want to have the shorthand use the logical tree and the non-shorthand use the visual? Or do we want to just use the visual tree and change the code so it can support an `AncestorType` of `null`? Or should we default it to `IControl`?

grokys commented:
Well for the longhand I think we're bound to follow WPF's syntax. The WPF syntax isn't confusing because you always have to supply the control type. But if for the shorthand we're allowing the user to _not_ supply a control type then I think it'd be confusing (and also not that useful) to use the visual tree.

So if that's accepted, then we need to support looking up in both the logical and visual tree.

If we need to support both then still my preferred solution is to supply a `public TreeType Tree { get; set; }` property on `RelativeSource`. The WPF longhand syntax would then default to `TreeType.Visual` and the shorthand syntax would use `TreeType.Logical`.

What do you think?

jkoritzinsky commented:
Sounds good to me! I'll implement it that way.

jkoritzinsky committed:
Updated implementation based on PR feedback.

jkoritzinsky committed:
Merge branch 'master' into RelativeSourceSyntaxSugar

jkoritzinsky committed:
Merge branch 'RelativeSourceSyntaxSugar' of https://github.com/jkoritzinsky/Avalonia into RelativeSourceSyntaxSugar

jkoritzinsky committed:
Clean up tests and fix PR feedback.

jkoritzinsky commented:
I've updated the implementation as per your comments. Can you take another look?

danwalmsley commented:
I know its a bit late, however allow me to suggest what I think could be a minor improvement for the syntax...

instead of
`
$parent[ns:Type; Level]
`

types could be declared in <> brackets like a generic class in c#?
so this would become 

`
$parent<ns:Type>[Level]
`

essentially [] is reserved for indexing, and <> is reserved for type selection, might be more memorable than [type;level] could easily forget the order?

jkoritzinsky commented:
We can't do angle brackets because that gets funky when the XAML is being parsed. And I don't want our users to have to escape the angle brackets. It would ruin the cleanness of the shorthand.

danwalmsley commented:
ah of course! :)

grokys commented on tests/Avalonia.Markup.Xaml.UnitTests/Xaml/BindingTests_RelativeSource.cs:
```diff
@@ -78,6 +78,87 @@ public void Binding_To_First_Ancestor_Works()
         }
 
         [Fact]
+        public void Binding_To_First_Ancestor_Without_AncestorType_Uses_LogicalTree()
```
Hmm, I still think that this implicit switching of trees depending on whether `AncestorType` is specified or not could be confusing. I'd prefer to require `AncestorType` when `Tree=Visual` (following WPF) and require `Tree=Logical` to switch over to the logical tree where `AncestorType` doesn't need to be specified. What do you think?

jkoritzinsky replied:
I'll make `AncestorType` be required when `Tree=Visual`. Sounds reasonable to me.

The automatic switch isn't happening any more. I just forgot to remove the test.

grokys commented on tests/Avalonia.Markup.Xaml.UnitTests/Xaml/BindingTests_RelativeSource.cs:
```diff
@@ -78,6 +78,87 @@ public void Binding_To_First_Ancestor_Works()
+
+                window.ApplyTemplate();
+
+                // This isn't needed in `Binding_To_First_Ancestor_Without_AncestorType_Uses_LogicalTree` -
```
I think we can remove this block.

grokys reviewed (changes requested):
@jkoritzinsky - _really_ sorry for the delay on reviewing this. I think we're just about there, but just one thing I'm not convinced about. I'm willing to be persuaded though :)

jkoritzinsky committed:
Make AncestorType required for ancestor searching via the visual tree.

jkoritzinsky committed:
Merge remote-tracking branch 'origin' into RelativeSourceSyntaxSugar

grokys commented on src/Markup/Avalonia.Markup.Xaml/Data/Binding.cs:
```diff
@@ -271,12 +222,13 @@ private ExpressionObserver CreateElementObserver(IControl target, string element
 
             return new ExpressionObserver(
-                ControlLocator.Track(target, RelativeSource.AncestorType, RelativeSource.AncestorLevel -1),
+                ControlLocator.Track(target, relativeSource.AncestorType != null ? relativeSource.Tree : TreeType.Logical, relativeSource.AncestorLevel - 1, relativeSource.AncestorType),
```
One final thing: is `relativeSource.AncestorType != null ? relativeSource.Tree : TreeType.Logical` still needed here now that the visual tree can't be searched with no `AncestorType`?

jkoritzinsky replied:
Good catch! I'll get rid of that.

jkoritzinsky committed:
Remove tree-switching on AncestorType == null

jkoritzinsky committed:
Merge branch 'master' into RelativeSourceSyntaxSugar

grokys committed:
Merge branch 'master' into RelativeSourceSyntaxSugar

grokys commented:
Merged! Thanks so much for this Jeremy!
